src/monitor/group_state_machine.c

```diff
@@ -295,14 +341,533 @@ ProceedGroupState(AutoFailoverNode *activeNode)
 			"Setting goal state of %s:%d to catchingup after it "
 			"converged to demotion and %s:%d converged to wait_primary.",
 			activeNode->nodeName, activeNode->nodePort,
-			otherNode->nodeName, otherNode->nodePort);
+			primaryNode->nodeName, primaryNode->nodePort);
 
 		/* it's safe to rejoin as a secondary */
 		AssignGoalState(activeNode, REPLICATION_STATE_CATCHINGUP, message);
 
 		return true;
 	}
 
+	/*
+	 * when a new primary is ready:
+	 *  report_lsn -> catchingup
+	 */
+	if (IsCurrentState(activeNode, REPLICATION_STATE_REPORT_LSN) &&
+		IsCurrentState(primaryNode, REPLICATION_STATE_PRIMARY))
+	{
+		char message[BUFSIZE];
+
+		LogAndNotifyMessage(
+			message, BUFSIZE,
+			"Setting goal state of %s:%d to secondary after %s:%d converged "
+			"to primary.",
+			activeNode->nodeName, activeNode->nodePort,
+			primaryNode->nodeName, primaryNode->nodePort);
+
+		/* it's safe to rejoin as a secondary */
+		AssignGoalState(activeNode, REPLICATION_STATE_SECONDARY, message);
+
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * Group State Machine when a primary node contacts the monitor.
+ */
+static bool
+ProceedGroupStateForPrimaryNode(AutoFailoverNode *primaryNode)
+{
+	List *otherNodesGroupList = AutoFailoverOtherNodesList(primaryNode);
+	int otherNodesCount = list_length(otherNodesGroupList);
+
+	/*
+	 * when a first "other" node wants to become standby:
+	 *  single -> wait_primary
+	 */
+	if (IsCurrentState(primaryNode, REPLICATION_STATE_SINGLE))
+	{
+		ListCell *nodeCell = NULL;
+
+		foreach(nodeCell, otherNodesGroupList)
+		{
+			AutoFailoverNode *otherNode = (AutoFailoverNode *) lfirst(nodeCell);
+
+			if (IsCurrentState(otherNode, REPLICATION_STATE_WAIT_STANDBY))
+			{
+				char message[BUFSIZE];
+
+				LogAndNotifyMessage(
+					message, BUFSIZE,
+					"Setting goal state of %s:%d to wait_primary after %s:%d "
+					"joined.", primaryNode->nodeName, primaryNode->nodePort,
+					otherNode->nodeName, otherNode->nodePort);
+
+				/* prepare replication slot and pg_hba.conf */
+				AssignGoalState(primaryNode,
+								REPLICATION_STATE_WAIT_PRIMARY,
+								message);
+
+				return true;
+			}
+		}
+	}
+
+	/*
+	 * when another node wants to become standby:
+	 *  primary -> join_primary
```

Comment:
Is failover disabled while the new standby is joining? If the primary fails during this time window, is that something we can recover from?

Reply:
We allow failover when in the new state JOIN_PRIMARY, that's why we have a new state here and don't re-use WAIT_PRIMARY.